### PR TITLE
fix(coverage): v8 to support source maps with multiple sources

### PR DESCRIPTION
**Pull request by AriPerkkio**
### Description

While setting up Vitest monorepo's coverage, noticed that V8 shows only the first file of bundled sources.

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [ ] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [x] Ideally, include a test that fails without this PR but passes with it.
- [x] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [x] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

### CODE REVIEW

**Comment by AriPerkkio** [comment]
Looks like bug in Vite's SSR transform:

```js
import { fileURLToPath } from "node:url";
import { createServer } from "vite";

const server = await createServer({
  configFile: false,
  root: fileURLToPath(new URL(".", import.meta.url)),
  server: { port: 1337 },
});

await server.listen();

const ssr = await server.transformRequest("./dist/bundle.js", { ssr: true });
const web = await server.transformRequest("./dist/bundle.js", { ssr: false });

console.log({ ssr: ssr.map.sources, web: web.map.sources });
// {
//   ssr: [ '../src/first.ts' ],
//   web: [ '../src/first.ts', '../src/second.ts' ]
// }

server.close();
```

**Comment by AriPerkkio** [comment]
V8 coverage works fine with https://github.com/vitejs/vite/pull/17677 applied locally. Generated the snapshots with that patch.

File: package.json [hunk 1]
```diff
@@ -61,7 +61,7 @@
     "rollup-plugin-license": "^3.5.2",
     "tsx": "^4.16.5",
     "typescript": "^5.5.4",
-    "vite": "^5.3.3",
+    "vite": "^5.4.0",
```
Comment (AriPerkkio) [inline package.json:64]: Needed for https://github.com/vitejs/vite/pull/17677

File: packages/coverage-v8/src/provider.ts
Comment (AriPerkkio) [inline packages/coverage-v8/src/provider.ts:1]: Verified that all our old tests pass with these changes when used with older Vite versions. ✅

File: packages/coverage-v8/src/provider.ts [hunk 2]
```diff
@@ -457,39 +457,44 @@ export class V8CoverageProvider extends BaseCoverageProvider implements Coverage
...
 
-    const sources = [url]
-    if (map.sources && map.sources[0] && !url.endsWith(map.sources[0])) {
-      sources[0] = new URL(map.sources[0], url).href
+    const sources = (map.sources || [])
+      .filter((source): source is NonNullable<typeof source> => !!source)
+      .map(source => new URL(source, url).href)
+
+    if (sources.length === 0) {
+      sources.push(url)
```
Comment (AriPerkkio) [inline packages/coverage-v8/src/provider.ts:485]: I'm not sure if this is even needed. If it's removed, all tests pass just fine. Maybe let's try removing this in future and leave it here for now.

File: packages/coverage-v8/src/provider.ts [hunk 3]
```diff
@@ -457,39 +457,44 @@ export class V8CoverageProvider extends BaseCoverageProvider implements Coverage
...
+        source: code || sourcesContent[0],
+        originalSource: sourcesContent[0],
       }
     }
 
-    const sources = [url]
-    if (map.sources && map.sources[0] && !url.endsWith(map.sources[0])) {
-      sources[0] = new URL(map.sources[0], url).href
+    const sources = (map.sources || [])
+      .filter((source): source is NonNullable<typeof source> => !!source)
```
Comment (sheremet-va) [inline packages/coverage-v8/src/provider.ts:481]: you can do `source != null` and you don't even need to have `source is` then :P

this was added in the latest TS version

**Review by sheremet-va — APPROVED** [review]

File: packages/coverage-v8/src/provider.ts (hunk 3 above)
Comment (AriPerkkio) [reply packages/coverage-v8/src/provider.ts]: That works, nice! I though it would detect `!!` as well, and thought we were using older TS version. 💯

Comment (sheremet-va) [reply packages/coverage-v8/src/provider.ts]: Yeah, unfortunately it doesn't work with `!!` because it's not an explicit assertion 😢

**Review by sheremet-va — APPROVED** [review]

**Pull request merged**